### ISSUE TITLE
Glassfish 4 JaxRS support

### ISSUE THREAD

**Issue by pxsfir** [jamesagnew/hapi-fhir#606]
Hello James,

My name is Pete Sfiridis Ive been working in the Health Care industry a number of years. Ive used HAPI HL7 many times as well as other commercial HL7 parsers.

I'm starting to use HAPI FHR JaxRS Server for a project, but I need to use Glassfish 4. Basically because of how the constructor is coded Glassfish 4 doesn't start up my Patient resource provider (extends AbstractJaxRsResourceProvider). Here's a short summary of why. Btw I do have a fix, see comments below. If you'd like me to contribute I would need access to the Git repo. I cant create a new Pull Request and not sure why.

When deploying to Glassfish4 a NullPointerException occurs in the 
    public  JaxRsMethodBindings(AbstractJaxRsProvider theProvider, Class<? extends AbstractJaxRsProvider> theProviderClass) {
        for (final Method m : ReflectionUtil.getDeclaredMethods(theProviderClass)) {
            final BaseMethodBinding<?> foundMethodBinding = BaseMethodBinding.bindMethod(m, theProvider.getFhirContext(), theProvider);
stacktrace
Severe:   Exception while loading the app
Severe:   Exception while loading the app : EJB Container initialization error
java.lang.NullPointerException
	at org.cdsframework.iis.integration.rs.service.__EJB31_Generated__PatientProvider__Intf____Bean__.getFhirContext(Unknown Source)
	at ca.uhn.fhir.jaxrs.server.util.JaxRsMethodBindings.<init>(JaxRsMethodBindings.java:58)
	at ca.uhn.fhir.jaxrs.server.util.JaxRsMethodBindings.getMethodBindings(JaxRsMethodBindings.java:140)
	at org.cdsframework.iis.integration.rs.service.AbstractRsResourceProvider.<init>(AbstractRsResourceProvider.java:91)

Its primarily due to the timing of when the Object in Glassfish has access to the FhirContext.

The fix involved adding a postConstructor annotation to AbstractJaxRsResourceProvider and moving the 
JaxRsMethodBindings.getMethodBindings call to the post postConstructor. I also had to make some other minor changes as well as change two of the test classes.

Let me know how to proceed with communicating the changeset. I tried creating a New Pull Request but couldn't. I could fork it, and I can add my change to the fork.

At this point let me know how I should proceed, I've cloned the Master repo and created a local branch but am unable to push it up.

**Comment by gteichrow** [comment]
Hey pxsfir, wow, seeing this myself here (GF 4.1.1).  I can see your workaround.  I'm wondering if it would be a decent permanent 'fix'?  I'm technically running Payara, but samething.

**Comment by jamesagnew** [comment]
This sounds like a logical fix to me. I'd definitely be interested in a pull request to this effect.

**Comment by gteichrow** [comment]
Does seem like a container lifecycle issue and it is timing sensitive I think (i.e, on a large war/ear, it might behave inconsistently).  Curious if there is another way to do this...  Looking at options that end with the same result @pxsfir has come up.

**Comment by pxsfir** [comment]
The @PostConstruct annotation should work on any J2EE container therefore
moving the code under this annotation should not create any different
behavior.
For now I just extended AbstractJaxRsProvider and created my own version.
Ideally it would be nice to roll this change in the core project.

public abstract class AbstractRsResourceProvider<R extends IBaseResource>
extends AbstractJaxRsProvider

**Comment by gteichrow** [comment]
Sounds like a plan.  I'll take your approach short term and wait for it to show up.  Cheers.

**Comment by gteichrow** [comment]
@pxsfir, not sure if you are using AbstractJaxRsPageProvider, but do you think it needs the same approach with a PostContruct() method?

**Comment by pxsfir** [comment]
Im not using as of yet. The issue I had was with this.

    /**
     * This post constructor is necessary so a J2EE container (e.g.,
Glassfish) can initialize the bindings at the appropriate time)
     */
    @PostConstruct
    protected void setUpPostConstruct() {
        if (theProviderClass == null) {
            theProviderClass = getClass();
        }
        theBindings = JaxRsMethodBindings.getMethodBindings(this,
theProviderClass);
    }

Basically theBindings werent initialized in the timing that they need to be
initialized in Glassfish so moving it to a @PostConstruct took care of the
issue.

**Comment by gteichrow** [comment]
@pxsfir, It does seem to be an issue with AbstractJaxRsPageProvider having the same symptom and I took a similar approach to what you've described adding a PostConstruct aspect and removing that code from the two constructors into that new method:

```
        @PostConstruct
        public void postContruct() {
	    try {
	        myBinding = new PageMethodBinding(getFhirContext(), PageProvider.class.getMethod("getPage"));
	    } catch (Exception e) {
	        throw new ca.uhn.fhir.context.ConfigurationException(e);
	    }      
        }
```

If you'd like to make this change and submit with your pull, go for it.  If you'd like me to, I'll do that.  
Cheers.
